Folder/main.py

```diff
@@ -0,0 +1,61 @@
+import requests
+import asset_data
+import email_notifier
+
+data_manager = asset_data.DataManager()
+data_manager.update_price_data()
+crypto_data = data_manager.crypto_data
+global_data = data_manager.global_data
+email_manager = email_notifier.EmailNotifier()
+users_data = email_manager.get_emails_data()
```

Comment:
probably want this in its own function or something, not globally

Reply:
Is it as simple as it seems, literally just put all of that into a function and define it so the code is cleaner? Because I only run all of those once on main.py so if I put it into a function, it just adds a couple lines of code to define it? I feel like I'm not getting something

Or do you mean separate it into two separate functions, one in each of asset_data.py and email_nofifier.py?

Reply:
Generally, functions are better because they can be Unit Tested. Unit Testing involves calling functions and potentially mocking some API calls. 

If I were doing this, I might make an App class
```
from asset_data import DataManager
from email_notifier import EmailNotifier

class App:
    coins = [
         ("bitcoin", "btc", "Bitcoin"), 
        ("ethereum", "eth", "Ethereum"), 
        ("solana", "sol", "Solana"), 
         ("dogecoin", "doge". "Dogecoin")
]

    def __init__(self):
        self.data_service = DataManager()
        self.email_service = EmailNotifier()
        
        self.data_service.update_price_data()
        print("App has initialized")

    def run(self):
         print("Starting run")
        some_global_data = self._format_global_data()
        
         mcap_percents_dict = {}
        for (full_coin, coin_ticker, user_coin) in self.coins:
              mcap_percents_dict[user_coin] = self._format_data(full_coin, coin_ticker)
       24_hour_label: str = self._get_up_down_label()
       
       email_and_messages: list[(str, str)] = self._process_users(mcap_percents_dict, something.... else... here)
       for (user_email, message) in emails:
            self.email_manager.do_the_thing(user_email, message)
        print(f"Sent emails to {",".join([email[0] for email in emails])}")
```

and then down below just call
```
app = App()
app.run() # XD
```

I think the main concepts here are  
* Logging to get status
* Reusing shared functions instead of creating duplicated code
    * Now you have 4 coins. imagine scaling to 100
* Creating functions that encapsualte 1 specific action
    * The nuance here is now you have to think about the inputs and outputs and 'what each function does'. Same as the Reusing shared functions thing. 

I kinda lost wind towards the middle as you can probably tell but I hope this helps a bit.

Reply:
I'm actually flabbergasted bamboozled dumbfoundastonished that you knocked that out in such little time lol but it all makes sense, especially your points about scaling and thought process when creating functions. I'm kind of there but still drafting some of my code line-by-line after coming up with the general idea as you can also prob tell. So this basically takes my entire main.py and turns it into an app/another class, interesting. Now I'm wondering if most/all main.py's follow the same format